relay-event-schema/src/protocol/nel.rs

```diff
@@ -0,0 +1,249 @@
+//! Contains definitions for the Network Error Logging (NEL) interface.
+//!
+//! NEL is a browser feature that allows reporting of failed network requests from the client side.
+//! W3C Editor's Draft: <https://w3c.github.io/network-error-logging/>
+//! MDN: <https://developer.mozilla.org/en-US/docs/Web/HTTP/Network_Error_Logging>
+use std::ops::Sub;
+
+use chrono::{DateTime, Duration, Utc};
+
+use crate::processor::ProcessValue;
+use crate::protocol::{
+    AsPair, Event, HeaderName, HeaderValue, Headers, LogEntry, PairList, Request, TagEntry, Tags,
+    Timestamp,
+};
+#[cfg(feature = "jsonschema")]
+use relay_jsonschema_derive::JsonSchema;
+use relay_protocol::{Annotated, Empty, FromValue, IntoValue};
+use thiserror::Error;
+
+/// The NEL parsing errors.
+#[derive(Debug, Error)]
+pub enum NelError {
+    /// Unexpected format.
+    #[error("unexpected format")]
+    InvalidNel,
+    /// Incoming Json is unparsable.
+    #[error("incoming json is unparsable")]
+    InvalidJson(#[from] serde_json::Error),
+}
+
+/// Generated network error report (NEL).
+#[derive(Clone, Debug, Default, PartialEq, Empty, FromValue, IntoValue, ProcessValue)]
+#[cfg_attr(feature = "jsonschema", derive(JsonSchema))]
+pub struct NelBody {
+    /// The elapsed number of milliseconds between the start of the resource
+    /// fetch and when it was completed or aborted by the user agent.
+    pub elapsed_time: Annotated<u64>,
+    /// HTTP method.
+    pub method: Annotated<String>,
+    /// If request failed, the phase of its network error. If request succeeded, "application".
+    pub phase: Annotated<String>,
+    /// The HTTP protocol and version.
+    pub protocol: Annotated<String>,
+    /// Request's referrer, as determined by the referrer policy associated with its client.
+    pub referrer: Annotated<String>,
+    /// The sampling rate.
+    pub sampling_fraction: Annotated<f64>,
+    /// The IP address of the server where the site is hosted.
+    pub server_ip: Annotated<String>,
+    /// HTTP status code.
+    pub status_code: Annotated<i64>,
+    /// If request failed, the type of its network error. If request succeeded, "ok".
+    #[metastructure(field = "type")]
+    pub ty: Annotated<String>,
+}
+
+/// Models the content of a NEL report.
+///
+/// NOTE: This is the structure used inside the Event (serialization is based on Annotated
+/// infrastructure). We also use a version of this structure to deserialize from raw JSON
+/// via serde.
+///
+/// See <https://w3c.github.io/network-error-logging/>
+#[derive(Clone, Debug, Default, PartialEq, Empty, FromValue, IntoValue, ProcessValue)]
+#[cfg_attr(feature = "jsonschema", derive(JsonSchema))]
+pub struct Nel {
```

Comment:
nit: See prefix comment above.

```suggestion
pub struct Report {
```